editoast/src/projects.rs

```diff
@@ -0,0 +1,186 @@
+    Clone, QueryableByName, Queryable, Debug, Serialize, Deserialize, AsChangeset, Identifiable,
+)]
+#[diesel(table_name = osrd_infra_project)]
+pub struct Project {
+    pub id: i64,
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub creation_date: NaiveDateTime,
+    pub last_modification: NaiveDateTime,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectData {
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectPatch {
+    pub name: Option<String>,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+impl Project {
+    pub fn create(data: ProjectData, conn: &mut PgConnection) -> Result<Project> {
+        match sql_query(
+        "INSERT INTO osrd_infra_project (name, description, objectives, funders, budget, image, creation_date,last_modification, tags)
+         VALUES ($1, $2, $3, $4, $5, NULL,CURRENT_TIMESTAMP, CURRENT_TIMESTAMP,$6)
+         RETURNING *",
+    )
+    .bind::<Text, _>(data.name)
+    .bind::<Nullable<Text>,_>(data.description)
+    .bind::<Nullable<Text>,_>(data.objectives)
+    .bind::<Nullable<Array<Text>>,_>(data.funders)
+    .bind::<Nullable<Integer>,_>(data.budget)
+    .bind::<Nullable<Array<Text>>,_>(data.tags)
+    .get_result::<Project>(conn)
+    {
+        Ok(project) => Ok(project),
+        Err(err) => Err(err.into()),
+    }
+    }
+
+    pub fn list(conn: &mut PgConnection, page: i64, page_size: i64) -> Vec<Project> {
+        let offset = page_size * (page - 1);
+        dsl::osrd_infra_project
+            .limit(page_size)
+            .offset(offset)
+            .load(conn)
+            .expect("List project query failed")
+    }
+
+    pub fn retrieve(conn: &mut PgConnection, project_id: i64) -> Result<Project> {
+        match dsl::osrd_infra_project.find(project_id).first(conn) {
+            Ok(project) => Ok(project),
+            Err(DieselError::NotFound) => Err(ProjectError::NotFound(project_id).into()),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    pub fn delete(project_id: i64, conn: &mut PgConnection) -> Result<()> {
+        match delete(dsl::osrd_infra_project.filter(dsl::id.eq(project_id))).execute(conn) {
+            Ok(1) => Ok(()),
+            Ok(_) => Err(ProjectError::NotFound(project_id).into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    pub fn update(data: ProjectPatch, conn: &mut PgConnection, project_id: i64) -> Result<Project> {
```

Comment:
Few things here:

- Check out the update documentation https://diesel.rs/guides/all-about-updates.html. Especially the `AsChangeset` section.
- You can adapt `ProjectPatch` using `AsChangeset` derive. This will simplify the code.
- You can add a default `now` value for the timestamp. 
- You can add `Option` on each field so diesel will update only filled information. (check the documentation for more info).